src/rdkafka_request.c

```diff
@@ -734,8 +734,12 @@ rd_kafka_handle_OffsetCommit (rd_kafka_t *rk,
         int i;
 	int actions;
 
-        if (err)
-		goto err;
+        if (err) {
+		if ((rk->rk_cgrp->rkcg_flags & RD_KAFKA_CGRP_F_TERMINATE) && err == RD_KAFKA_RESP_ERR__TRANSPORT)
```

Comment:
This will prevent retries for the final offset commits, which is something we definitely want, so this will not work.

Reply:
consumer_close() will block up to session.timeout.ms for the final offset commit to finish.

Reply:
> This will prevent retries for the final offset commits, which is something we definitely want, so this will not work.

It would keep retrying (to send offset commits) infinitely (because `incr_retry` is 0), say if the brokers are down, then the consumer could never quit  `rdkafka_consumer_close`. It makes no sense.

Reply:
> consumer_close() will block up to session.timeout.ms for the final offset commit to finish.

No, it would block there forever (while the brokers are down), even the `session.timeout.ms` expires. Only while we let the OffsetCommitRequest response with failure (instead of infinite retrying), the `rkcg_wait_commit_cnt` could get a chance to be `0`, thus trigger the `TERMINATE` event, and then `rdkafka_consumer_close` could quit.

Reply:
> This will prevent retries for the final offset commits, which is something we definitely want, so this will not work.

While the network is down (`RD_KAFKA_RESP_ERR__TRANSPORT`) and the consumer is trying to close, why insist waiting for the network/cluster recovery to commit the offsets? Anyway, we get the OffsetCommit callback and we could get the offsets-commit-error, right?

Reply:
Great analysis of the code!

We can't bail out on ERR__TRANSPORT, the contract is for the consumer to commit its final offsets within the remaining session timeout, and it should retry even if the broker connection goes down. But you're right in that retrying the rkbuf will not work, unless the coordinator comes back up on the same broker.
So I'm suggesting a slightly modified fix where the request fails and the offset commit is put on the cgrp's deferred commit list: this list makes sure the commit is retried when the coordinator is back up, or that the commit is properly timed out after session.timeout.ms.

Please see https://github.com/edenhill/librdkafka/pull/2440
It would be great if you could try it out.

Reply:
I just had a try, -- all our testcases could pass. Cheers! :-)